tokio/src/runtime/metrics/histogram.rs

```diff
@@ -112,52 +224,62 @@ impl HistogramBatch {
     }
 
     fn value_to_bucket(&self, value: u64) -> usize {
-        match self.scale {
-            HistogramScale::Linear => {
-                let max = self.buckets.len() - 1;
-                cmp::min(value / self.resolution, max as u64) as usize
-            }
-            HistogramScale::Log => {
-                let max = self.buckets.len() - 1;
-
-                if value < self.resolution {
-                    0
-                } else {
-                    let significant_digits = 64 - value.leading_zeros();
-                    let bucket_digits = 64 - (self.resolution - 1).leading_zeros();
-                    cmp::min(significant_digits as usize - bucket_digits as usize, max)
-                }
-            }
-        }
+        self.configuration.value_to_bucket(value)
     }
 }
 
 impl HistogramBuilder {
     pub(crate) fn new() -> HistogramBuilder {
         HistogramBuilder {
-            scale: HistogramScale::Linear,
-            // Resolution is in nanoseconds.
-            resolution: 100_000,
-            num_buckets: 10,
+            histogram_type: HistogramType::Linear(LinearHistogram {
+                num_buckets: 10,
+                bucket_width: 100_000,
+            }),
+            legacy: None,
+        }
+    }
+
+    pub(crate) fn legacy_mut(&mut self, f: impl Fn(&mut LegacyBuilder)) {
+        if let Some(legacy) = &mut self.legacy {
+            f(legacy)
+        } else {
+            let mut legacy = LegacyBuilder::default();
+            f(&mut legacy);
+            self.legacy = Some(legacy)
         }
     }
 
     pub(crate) fn build(&self) -> Histogram {
-        let mut resolution = self.resolution;
+        let configuration = match &self.legacy {
+            Some(legacy) => {
+                let mut resolution = legacy.resolution;
 
-        assert!(resolution > 0);
+                assert!(resolution > 0);
 
-        if matches!(self.scale, HistogramScale::Log) {
-            resolution = resolution.next_power_of_two();
-        }
+                if matches!(legacy.scale, HistogramScale::Log) {
+                    resolution = resolution.next_power_of_two();
+                }
+                match legacy.scale {
+                    HistogramScale::Linear => HistogramType::Linear(LinearHistogram {
+                        num_buckets: legacy.num_buckets,
+                        bucket_width: resolution,
+                    }),
+                    HistogramScale::Log => HistogramType::LogLegacy(LegacyLogHistogram {
+                        num_buckets: legacy.num_buckets,
+                        first_bucket_width: resolution,
+                    }),
+                }
```

Comment:
Could these match statements be combined? We don't seem to use resolution after the second one. I think perhaps we wouldn't need the `mut` on `resolution` after that either.

```suggestion
                match legacy.scale {
                    HistogramScale::Linear => HistogramType::Linear(LinearHistogram {
                        num_buckets: legacy.num_buckets,
                        bucket_width: resolution,
                    }),
                    HistogramScale::Log => HistogramType::LogLegacy(LegacyLogHistogram {
                        num_buckets: legacy.num_buckets,
                        first_bucket_width: resolution.next_power_of_two(),
                    }),
                }
```